src/prefect/tasks/shell.py

```diff
@@ -101,26 +119,34 @@ def run(self, command: str = None, env: dict = None) -> str:
                 [self.shell, tmp.name], stdout=PIPE, stderr=STDOUT, env=current_env
             ) as sub_process:
                 lines = []
-                line = None
                 for raw_line in iter(sub_process.stdout.readline, b""):
                     line = raw_line.decode("utf-8").rstrip()
-                    if self.return_all:
-                        lines.append(line)
-                    else:
-                        # if we're returning all, we don't log every line
+                    lines.append(line)
+
+                    if self.stream_output:
                         self.logger.debug(line)
+
                 sub_process.wait()
                 if sub_process.returncode:
                     msg = "Command failed with exit code {}".format(
                         sub_process.returncode,
                     )
+
                     self.logger.error(msg)
 
-                    if self.log_stderr:
+                    if self.log_stderr and not self.stream_output:
                         self.logger.error("\n".join(lines))
 
-                    raise prefect.engine.signals.FAIL(msg) from None  # type: ignore
-        if self.return_all:
-            return lines
-        else:
-            return line
+                    # Return a exit code and lines so the failure can be inspected
+                    raise prefect.engine.signals.FAIL(
+                        msg,
+                        result=ShellResult(
```

Comment:
I'm not sure I follow the need have a `ShellResult` typed dict as opposed to a standard dictionaty :thinking:

I also think this functionality of attaching the lines should be gated behind some kwarg like `return_stdout_on_error` (probably a better name than this) and keep with only attaching the error code and message. As an example on why I think it should be gated: I recall some users performing terraform commands with this task and if there's an error there would be quite a bit of lines returned in the output.

Reply:
Well I think just returning a dictionary without information about what's in it isn't a pattern I like. I'd prefer to have the return type of the function be meaningful (hence a typed dict or a lightweight class).

Are the lines being returned in the signal result displayed or something? Wouldn't they just not access them if they don't need them?

Reply:
I don't feel strongly about it so no preference but a dictionary with static keys

```python
{
  output: [],
  return_code: ...
}
```

that are documented in the docstring, to me, feels like is provides the same benefit.

Thinking on this a bit more I actually am more in favor of also using the `return_all` logic here (instead of the kwarg to gate it) that is present at the end of the task. i.e. if I don't opt in to returning all of my lines then I am expecting the FAIL signal's return value to only have the last line.

Reply:
It's missing the benefit of:
- Understanding the function's return by looking at the type signature
- mypy type checking of the return value and access to its members
- Enforcing and indicating types of the members of the dict

`return_all` makes sense to me! I'll implement that.

Reply:
I think josh is right here. Since the `ShellResult` is only part of a raised `FAIL` signal, a dict here is best. We need it json-serializable, and users are unlikely to deal with it programmatically (and if they do, documenting the fields in the docstring makes the most sense to me).